utils/controller/controller.go

```diff
@@ -18,6 +23,72 @@ import (
 	logutil "github.com/argoproj/argo-rollouts/utils/log"
 )
 
+// processNextWatchObj will process a single object from the watch by seeing if
+// that object is in an index and enqueueing the value object from the indexer
+func processNextWatchObj(watchEvent watch.Event, queue workqueue.RateLimitingInterface, indexer cache.Indexer, index string) {
+	obj := watchEvent.Object
+	acc, err := meta.Accessor(obj)
+	if err != nil {
+		log.Errorf("Error processing object from watch: %v", err)
+		return
+	}
+	objsToEnqueue, err := indexer.ByIndex(index, fmt.Sprintf("%s/%s", acc.GetNamespace(), acc.GetName()))
+	if err != nil {
+		log.Errorf("Cannot process indexer: %s", err.Error())
+		return
+	}
+	for i := range objsToEnqueue {
+		Enqueue(objsToEnqueue[i], queue)
+	}
+}
+
+// WatchResource is a long-running function that will continually call the
+// processNextWatchObj function in order to watch changes on a resource kind
+// and enqueue a different resources kind that interact with them
+func WatchResource(client dynamic.Interface, namespace string, gvk schema.GroupVersionResource, queue workqueue.RateLimitingInterface, indexer cache.Indexer, index string) error {
+	log.Infof("Starting watch on resource '%s'", gvk.Resource)
+	var watchI watch.Interface
+	var err error
+	if namespace == metav1.NamespaceAll {
+		watchI, err = client.Resource(gvk).Watch(metav1.ListOptions{})
+	} else {
+		watchI, err = client.Resource(gvk).Namespace(namespace).Watch(metav1.ListOptions{})
+	}
+
+	if err != nil {
+		log.Errorf("Error with watch: %v", err)
+		return err
+	}
+	for watchEvent := range watchI.ResultChan() {
+		processNextWatchObj(watchEvent, queue, indexer, index)
+	}
+	return nil
```

Comment:
There should be a `watchI.Stop()` call which will clean up resources held by the watch.